other/word_search.py

```diff
@@ -0,0 +1,395 @@
+            for col in cols:
+                # Check if there is space to the right of the word as well as above
+                if word_length > col + 1:
+                    continue
+
+                # Check if there are existing letters
+                # to the right of the column that will be overwritten
+                letters_diagonal_left = [
+                    self.board[row - i][col - i] for i in range(word_length)
+                ]
+                if all(letter is None for letter in letters_diagonal_left):
+                    # Successful, insert the word north
+                    for i in range(word_length):
+                        self.board[row - i][col - i] = word[i]
+                    return
+
+    def generate_board(self) -> None:
+        """
+        Generates a board with a random direction for each word.
+
+        >>> wt = WordSearch(WORDS, WIDTH, HEIGHT)
+        >>> wt.generate_board()
+        >>> len(list(filter(lambda word: word is not None, sum(wt.board, start=[])))
+        ... ) == sum(map(lambda word: len(word), WORDS))
+        True
+        """
+        directions = (
+            self.insert_north,
+            self.insert_north_east,
+            self.insert_east,
+            self.insert_south_east,
+            self.insert_south,
+            self.insert_south_west,
+            self.insert_west,
+        )
+        for word in self.words:
+            # Shuffle the row order and column order that is used when brute forcing
+            # the insertion of the word
+            rows, cols = list(range(self.height)), list(range(self.width))
+            shuffle(rows)
+            shuffle(cols)
+
+            # Insert the word via the direction
+            choice(directions)(word, rows, cols)
+
+
+def visualise_word_search(
+    board: list[list[str | None]] | None = None, *, add_fake_chars: bool = True
+) -> None:
+    """
+    Graphically displays the word search in the terminal.
+
+    >>> ws = WordSearch(WORDS, 5, 5)
+    >>> ws.insert_north("cat", [4], [4])
+    >>> visualise_word_search(
+    ...     ws.board, add_fake_chars=False)  # doctest: +NORMALIZE_WHITESPACE
+    # # # # #
+    # # # # #
+    # # # # t
+    # # # # a
+    # # # # c
+    >>> ws.insert_north_east("snake", [4], [4, 3, 2, 1, 0])
+    >>> visualise_word_search(
+    ...     ws.board, add_fake_chars=False)  # doctest: +NORMALIZE_WHITESPACE
+    # # # # e
+    # # # k #
+    # # a # t
+    # n # # a
+    s # # # c
+    """
+    if board is None:
+        word_search = WordSearch(WORDS, WIDTH, HEIGHT)
+        word_search.generate_board()
+        board = word_search.board
+
+    result = ""
+    for row in range(len(board)):
+        for col in range(len(board[0])):
+            character = "#"
+            if (letter := board[row][col]) is not None:
+                character = letter
+            # Empty char, so add a fake char
+            elif add_fake_chars:
+                character = chr(randint(97, 122))
```

Comment:
Can we have input words restricted to just lowercase English letters? If you're just padding out the word search with random English lowercase letters and the user chooses words that use other characters (uppercase English letters, etc) then the inserted words would stand out.

Reply:
We can limit to only lowercase english letters 👍